libs/vkd3d/device.c

```diff
@@ -5076,6 +5118,208 @@ struct d3d12_device *unsafe_impl_from_ID3D12Device(d3d12_device_iface *iface)
+    TRACE("iface %p, handle %p \n", iface, handle);
+    if(!iface || !handle)
+        return E_INVALIDARG;
+    
+    device = d3d12_device_from_ID3D12DeviceExt(iface);
+    vk_device = device->vk_device;
+    vk_procs = &device->vk_procs;
+
+    VK_CALL(vkDestroyCuFunctionNVX(vk_device, handle->vkCuFunction, NULL));
+    VK_CALL(vkDestroyCuModuleNVX(vk_device, handle->vkCuModule, NULL));
+    vkd3d_free(handle);
+    return S_OK;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_vkd3d_ext_GetCudaTextureObject(ID3D12DeviceExt *iface, D3D12_CPU_DESCRIPTOR_HANDLE srv_handle,
+       D3D12_CPU_DESCRIPTOR_HANDLE sampler_handle, UINT32 *cuda_texture_handle)
+{
+    VkImageViewHandleInfoNVX imageViewHandleInfo = { VK_STRUCTURE_TYPE_IMAGE_VIEW_HANDLE_INFO_NVX };
+    const struct vkd3d_vk_device_procs *vk_procs;
+    struct d3d12_desc *sampler_desc;
+    struct d3d12_device *device;
+    struct d3d12_desc *srv_desc;
+    
+    TRACE("iface %p, srv_handle %#x, sampler_handle %#x, cuda_texture_handle %p.\n", iface, srv_handle, sampler_handle, cuda_texture_handle);
+    if(!cuda_texture_handle)
+       return E_INVALIDARG;
+
+    device = d3d12_device_from_ID3D12DeviceExt(iface);
+    srv_desc = d3d12_desc_from_cpu_handle(srv_handle);
+    sampler_desc = d3d12_desc_from_cpu_handle(sampler_handle);
+
+    imageViewHandleInfo.imageView = srv_desc->info.view->vk_image_view;
+    imageViewHandleInfo.sampler = sampler_desc->info.view->vk_sampler;
+    imageViewHandleInfo.descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER;
+
+    vk_procs = &device->vk_procs;
+    *cuda_texture_handle = VK_CALL(vkGetImageViewHandleNVX(device->vk_device, &imageViewHandleInfo));
+    return S_OK;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_vkd3d_ext_GetCudaSurfaceObject(ID3D12DeviceExt *iface, D3D12_CPU_DESCRIPTOR_HANDLE uav_handle, 
+        UINT32 *cuda_surface_handle)
+{
+    VkImageViewHandleInfoNVX imageViewHandleInfo = { VK_STRUCTURE_TYPE_IMAGE_VIEW_HANDLE_INFO_NVX };
+    const struct vkd3d_vk_device_procs *vk_procs;
+    struct d3d12_device *device;
+    struct d3d12_desc *uav_desc;
+    
+    TRACE("iface %p, uav_handle %#x, cuda_surface_handle %p.\n", iface, uav_handle, cuda_surface_handle);
+    if(!cuda_surface_handle)
+       return E_INVALIDARG;
+
+    device = d3d12_device_from_ID3D12DeviceExt(iface);
+    uav_desc = d3d12_desc_from_cpu_handle(uav_handle);
+
+    imageViewHandleInfo.imageView = uav_desc->info.view->vk_image_view;
+    imageViewHandleInfo.descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_IMAGE;
+
+    vk_procs = &device->vk_procs;
+    *cuda_surface_handle = VK_CALL(vkGetImageViewHandleNVX(device->vk_device, &imageViewHandleInfo));
+    return S_OK; 
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_vkd3d_ext_CaptureUAVInfo(ID3D12DeviceExt *iface, D3D12_UAV_INFO *uav_info)
+{
+    if(!uav_info)
+       return E_INVALIDARG;
+
+    TRACE("iface %p, uav_info %p.\n", iface, uav_info);
+#if defined(_WIN32)
+    d3d12_uav_info = uav_info;
```

Comment:
??

Reply:
I'm not sure what this aims to accomplish. If the goal is to create a UAV and get the ViewHandle, couldn't this be a new interface entry point, similar to the SRV/Sampler path?

Reply:
CaptureUAVInfo() supposed to capture the information from the next CreateUnorderedAccess() on the same thread.

Reply:
But why is this desired? This is a strange interface decision since you could just use GetCudaSurface() already either way.

Reply:
I agree this is strange.! However, we cannot change the behavior since DLSS is shipped with these APIs.

Reply:
._. well, if we have no choice ... At least add comments explaining why this API wart is required.